tests/zfs-tests/tests/functional/zap_shrink/zap_shrink_001_pos.ksh

```diff
@@ -0,0 +1,79 @@
+#! /bin/ksh -p
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or https://opensource.org/licenses/CDDL-1.0.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright 2024, Klara Inc.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# Create a large number of files in a directory. Then remove all files and
+# check that the directory zap was shrunk. Use zdb to check that the zap object
+# contains only one leaf block using zdb.
+#
+
+verify_runnable "global"
+
+DIR=largedir
+
+NR_FILES=100000
+BATCH=1000
+CWD=$PWD
+
+log_assert "Create a large number of files ($NR_FILES) in a directory. " \
+	"Make sure that the directory ZAP object was shrunk."
+
+log_must mkdir $TESTDIR/$DIR
+
+cd $TESTDIR/$DIR
+# In order to prevent arguments overflowing, create NR_FILES in BATCH at once.
+for i in $(seq $(($NR_FILES/$BATCH))); do
+	touch $(seq $((($i-1)*$BATCH+1)) $(($i*$BATCH)));
+done
+cd $CWD
+
+log_must test $NR_FILES -eq $(ls -U $TESTDIR/$DIR | wc -l)
+
+# remove all files in $DIR directory
+cd $TESTDIR/$DIR
+for i in $(seq $(($NR_FILES/$BATCH))); do
+	rm $(seq $((($i-1)*$BATCH+1)) $(($i*$BATCH)))
+done
+cd $CWD
+sync_pool $TESTPOOL
+
+log_must test 0 -eq $(ls -U $TESTDIR/$DIR | wc -l)
+
+# check whether zap_shrink works
+zapobj=$(zdb -v -O $TESTPOOL/$TESTFS $DIR)
+nleafs=$(echo "$zapobj" | grep "Leaf blocks:" | awk -F\: '{print($2);}')
+log_must test 1 -eq $nleafs
+
+log_must zpool export $TESTPOOL
+zapobj=$(zdb -e -v -O $TESTPOOL/$TESTFS $DIR)
```

Comment:
Hmm, it's possible `-O` doesn't work with an exported pool.  If that's the case we can skip the export/import steps like above run `zdb` on the imported pool.  Let's just make sure to `sync_pool $TESTPOOL` before calling `zdb` to ensure everything's on disk.

```
Test: /usr/share/zfs/zfs-tests/tests/functional/zap_shrink/zap_shrink_001_pos (run as root) [00:34] [FAIL]
22:26:09.27 ASSERTION: Create a large number of files (100000) in a directory.  Make sure that the directory ZAP object was shrunk.
22:26:09.28 SUCCESS: mkdir /mnt/testdir/largedir
22:26:26.30 SUCCESS: test 100000 -eq 100000
22:26:43.77 SUCCESS: zpool sync testpool
22:26:43.78 SUCCESS: test 0 -eq 0
22:26:43.82 SUCCESS: test 1 -eq 1
22:26:43.88 SUCCESS: zpool export testpool
22:26:43.95 failed to hold dataset 'testpool/testfs': No such file or directory
22:26:43.95 /usr/share/zfs/zfs-tests/tests/functional/zap_shrink/zap_shrink_001_pos.ksh[76]: log_must[70]: log_pos[265]: test: argument expected
22:26:43.95 ERROR: test 1 -eq exited 2
```